File: cli/ts/charts/ticker.ts

import ora from 'ora';
import blessed from 'blessed';
import { loadSettingsAndCreateExchange } from '../helpers.js';

const spinner = ora ('Loading Ticker...');

async function plotTicker (exchangeNames: string, symbol: string, args: any) {
    spinner.start ();
    const exchangeIds = exchangeNames.split (',');

    const exchangePromises = exchangeIds.map ((id) => loadSettingsAndCreateExchange (id, {}));

    const exchanges = await Promise.all (exchangePromises);

    const boxWidth = 32;
    const boxHeight = 10 * 2 + 5;

    const screen = blessed.screen ({
        'smartCSR': true,
        'title': `Ticker - ${symbol}`,
    });

    // append screens
    const boxes = [];
    for (let i = 0; i < exchanges.length; i++) {
        const box = blessed.box ({
            'top': 'center',
            'left': boxWidth * i + 2,
            'width': boxWidth,
            'height': boxHeight,
            'tags': true,
            'border': {
                'type': 'line',
            },
            'style': {
                'fg': 'white',
                'bg': 'black',
                'border': { 'fg': 'magenta' },
            },
        });
        boxes.push (box);
        screen.append (box);
    }
    screen.key ([ 'escape', 'q', 'C-c' ], () => {
        process.exit (0);
    });

    process.on ('SIGINT', () => {
        screen.destroy (); // cleanup blessed screen
        spinner.clear ();
        process.exit (0);
    });
    // screen.render ();
    const renderPromises = [];
    for (let i = 0; i < exchanges.length; i++) {
        renderPromises.push (renderTicker (screen, exchanges[i], symbol, boxes[i], exchanges[i].id));
    }
    await Promise.all (renderPromises);
}

function center (str, width) {
    const len = str.length;
    const pad = Math.max (0, width - len);
    const left = Math.floor (pad / 2);
    const right = pad - left;
    return ' '.repeat (left) + str + ' '.repeat (right);
}

async function renderTicker (screen, exchange, symbol, box, name) {
    try {
        while (true) {
            const ticker = await exchange.watchTicker (symbol);
            if (spinner.isSpinning) {
                spinner.clear ();
                spinner.stop ();
            }
            const {
                last,
                percentage,
                change,
                open,
                high,
                low,
                baseVolume,
                quoteVolume,
                bid,
                ask,
                vwap,
                average,
                previousClose,
                indexPrice,
                markPrice,
                bidVolume,
                askVolume,
                datetime,
            } = ticker;

            function fmt (value, digits = 2) {
                return (typeof value === 'number') ? value.toFixed (digits) : 'undefined';
            }

            function fmtDate (datetime) {
                if (!datetime) {
                    return undefined;
                }
                return datetime.split ('.')[0];
            }

            function colorize (label, value, color = 'white') {
                return `{bold}${center (label, 12)}:{/bold} {${color}-fg}${center (value, 14)}{/${color}-fg}`;
            }

            function addLine (content: string) {
                return content;
                // content += '{center}{gray-fg}─────────────────────{/gray-fg}{/center}\n';
                const line = '─'.repeat (33); // 35 - 2 (for borders)
                content += `{center}{gray-fg}${line}{/gray-fg}{/center}\n`;
                return content;
            }

            function colorForNumber (val) {
                if (typeof val !== 'number') return 'white';
                if (val > 0) return 'green';
                if (val < 0) return 'red';
                return 'white';
            }

            let content = `\n{center}{bold}${name} — ${symbol}{/bold}{/center}\n\n\n`;

            content += colorize ('Last', fmt (last)) + '\n';
            content = addLine (content);
            content += colorize ('Change', fmt (change), colorForNumber (change)) + '\n';
            content = addLine (content);
            content += colorize ('Percent', percentage !== undefined ? percentage.toFixed (2) + '%' : 'undefined', 'cyan') + '\n';
            content = addLine (content);
            content += colorize ('Open', fmt (open)) + '\n';
            content = addLine (content);
            content += colorize ('High', fmt (high)) + '\n';
            content = addLine (content);
            content += colorize ('Low', fmt (low)) + '\n';
            content = addLine (content);
            content += colorize ('Bid', fmt (bid), 'green') + '\n';
            content = addLine (content);
            content += colorize ('Ask', fmt (ask), 'red') + '\n';
            content = addLine (content);
            content += colorize ('Base Vol', fmt (baseVolume)) + '\n';
            content = addLine (content);
            content += colorize ('Quote Vol', fmt (quoteVolume)) + '\n';
            content = addLine (content);
            content += colorize ('VWAP', fmt (vwap)) + '\n';
            content = addLine (content);
            content += colorize ('Avg', fmt (average), 'yellow') + '\n';
            content = addLine (content);
            content += colorize ('Prev Close', fmt (previousClose)) + '\n';

            content = addLine (content);
            content += colorize ('Index Price', fmt (indexPrice)) + '\n';

            content = addLine (content);
            content += colorize ('Mark Price', fmt (markPrice)) + '\n';

            content = addLine (content);
            content += colorize ('Bid Vol', fmt (bidVolume)) + '\n';

            content = addLine (content);
            content += colorize ('Ask Vol', fmt (askVolume)) + '\n';

            content = addLine (content);
            content += colorize ('Datetime', formatShortDate (datetime), 'cyan') + '\n';

            box.setContent (content);
            screen.render ();
        }
    } catch (err) {
        box.setContent (`{red-fg}Error: ${err.message}{/red-fg}`);
        screen.render ();
        setTimeout (() => renderTicker (screen, exchange, symbol, box, name), 3000);
    }
}

function formatShortDate (input) {
    if (!input) {
        return undefined;
    }
    // Fill in missing seconds if needed
    let normalized = input;
    if (input.endsWith (':')) {
        normalized += '01';
    }

    const date = new Date (normalized);
    const pad = (n) => String (n).padStart (2, '0');

    const month = pad (date.getMonth () + 1);
    const day = pad (date.getDate ());
    const hours = pad (date.getHours ());
    const minutes = pad (date.getMinutes ());
    const seconds = date.getSeconds (); // no padding

    return `${month}/${day} ${hours}:${minutes}:${seconds}`;
}

export {
    plotTicker,
};

File: java/lib/src/main/java/io/github/ccxt/types/Ticker.java
package io.github.ccxt.types;

import java.util.Map;

public final class Ticker {
    public String symbol;
    public Long timestamp;
    public String datetime;
    public Double high;
    public Double low;
    public Double bid;
    public Double bidVolume;
    public Double ask;
    public Double askVolume;
    public Double vwap;
    public Double open;
    public Double close;
    public Double last;
    public Double previousClose;
    public Double change;
    public Double percentage;
    public Double average;
    public Double quoteVolume;
    public Double baseVolume;
    public Double indexPrice;
    public Double markPrice;
    public Map<String, Object> info;

    @SuppressWarnings("unchecked")
    public Ticker(Object raw) {
        Map<String, Object> data = TypeHelper.toMap(raw);
        this.symbol = TypeHelper.safeString(data, "symbol");
        this.timestamp = TypeHelper.safeInteger(data, "timestamp");
        this.datetime = TypeHelper.safeString(data, "datetime");
        this.high = TypeHelper.safeFloat(data, "high");
        this.low = TypeHelper.safeFloat(data, "low");
        this.bid = TypeHelper.safeFloat(data, "bid");
        this.bidVolume = TypeHelper.safeFloat(data, "bidVolume");
        this.ask = TypeHelper.safeFloat(data, "ask");
        this.askVolume = TypeHelper.safeFloat(data, "askVolume");
        this.vwap = TypeHelper.safeFloat(data, "vwap");
        this.open = TypeHelper.safeFloat(data, "open");
        this.close = TypeHelper.safeFloat(data, "close");
        this.last = TypeHelper.safeFloat(data, "last");
        this.previousClose = TypeHelper.safeFloat(data, "previousClose");
        this.change = TypeHelper.safeFloat(data, "change");
        this.percentage = TypeHelper.safeFloat(data, "percentage");
        this.average = TypeHelper.safeFloat(data, "average");
        this.quoteVolume = TypeHelper.safeFloat(data, "quoteVolume");
        this.baseVolume = TypeHelper.safeFloat(data, "baseVolume");
        this.indexPrice = TypeHelper.safeFloat(data, "indexPrice");
        this.markPrice = TypeHelper.safeFloat(data, "markPrice");
        this.info = TypeHelper.getInfo(data);
    }
}
